### DISCUSSION TITLE
Ability to assert that an included relationship cannot be null?

### DISCUSSION THREAD

**Discussion by olalonde** [prisma/prisma#20987] [Q&A]
### Question

I have a query that looks like this:

```typescript
  const users = await prisma.user.findMany({
    where: {
      // Use AND to ensure all conditions are met
      AND: [
        {
          NOT: {
            imAccount: null,
          },
        },
      ],
    },
    include: {
      imAccount: true,
    },
    take: limit,
  });
```

The return type still says `imAccount` is possibly null (although this is not possible due to the where clause. Is there any way to work around this?

**Comment by olalonde** [comment]
Possible workaround?

```typescript
type NonNullableArrayProps<T, P extends string> = {
  [K in keyof T]: T[K] extends { [Q in P]: infer U }
    ? U extends null
      ? never // Exclude the type if it is null
      : T[K] & { [Q in P]: U }
    : T[K];
};

 const users = await prisma.user.findMany({
    where: {
      // Use AND to ensure all conditions are met
      AND: [
        {
          NOT: {
            imAccount: null,
          },
        },
      ],
    },
    include: {
      imAccount: true,
    },
    take: limit,
  }) as NonNullableArrayProps<typeof users, "imAccount">;
```

Didn't test extensively...

**Reply by ludralph** [reply]
Hi @olalonde 👋 

I think the issue here is that TypeScript is not able to infer that the `imAccount` field will never be null based on the where clause in your Prisma query. Unfortunately, there's no direct way to instruct TypeScript about this inference based on the Prisma query. TypeScript's type system doesn't have a built-in way to understand the implications of your runtime logic (the where clause in this case).
As a workaround, you can use a type assertion to tell TypeScript that `imAccount` will never be null just like the workaround you provided.
```ts
const users = (await prisma.user.findMany({
  where: {
    // Use AND to ensure all conditions are met
    AND: [
      {
        NOT: {
          imAccount: null,
        },
      },
    ],
  },
  include: {
    imAccount: true,
  },
  take: limit,
})) as (User & { imAccount: ImAccount })[]; // Replace User and ImAccount with your actual types
```

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
